progressutil/iocopy.go

```diff
@@ -93,69 +97,56 @@ func (cpp *CopyProgressPrinter) AddCopy(reader io.Reader, name string, size int6
 	cr.pb.SetPrintAfter(cr.formattedProgress())
 
 	cpp.readers = append(cpp.readers, cr)
-	cpp.lock.Unlock()
 
 	go func() {
 		_, err := io.Copy(dest, cr)
-		if err != nil {
-			cpp.lock.Lock()
-			cpp.errors = append(cpp.errors, err)
-			cpp.lock.Unlock()
-		}
-		cr.setDone(true)
+		cpp.results <- err
 	}()
+	return nil
 }
 
 // PrintAndWait will print the progress for each copy operation added with
 // AddCopy to printTo every printInterval. This will continue until every added
 // copy is finished, or until cancel is written to.
+// PrintAndWait may only be called once; any subsequent calls will immediately
+// return ErrAlreadyStarted.  After PrintAndWait has been called, no more
+// copies may be added to the CopyProgressPrinter.
 func (cpp *CopyProgressPrinter) PrintAndWait(printTo io.Writer, printInterval time.Duration, cancel chan struct{}) error {
-	for {
-		// If cancel is not nil, see if anything has been written to it. If
-		// something has, return, otherwise keep drawing.
-		if cancel != nil {
-			select {
-			case <-cancel:
-				return nil
-			default:
-			}
-		}
-
-		cpp.lock.Lock()
-		readers := cpp.readers
-		errors := cpp.errors
+	cpp.lock.Lock()
+	if cpp.started {
 		cpp.lock.Unlock()
+		return ErrAlreadyStarted
+	}
+	cpp.started = true
+	cpp.lock.Unlock()
 
-		if len(errors) > 0 {
-			return errors[0]
-		}
+	n := len(cpp.readers)
+	if n == 0 {
+		// Nothing to do.
+		return nil
+	}
 
-		if len(readers) > 0 {
+	t := time.NewTicker(printInterval)
+	for i := 0; i < n; {
+		select {
+		case <-cancel:
+			return nil
+		case <-t.C:
 			_, err := cpp.pbp.Print(printTo)
 			if err != nil {
 				return err
 			}
-		}
-
-		allDone := true
-		for _, r := range readers {
-			allDone = allDone && r.getDone()
-		}
-		if allDone && len(readers) > 0 {
-			// We still need to check for errors again, as one may
-			// have occurred between when we first checked and when
-			// the readers were marked as done
-			cpp.lock.Lock()
-			errors = cpp.errors
-			cpp.lock.Unlock()
-			if len(errors) > 0 {
-				return errors[0]
+		case err := <-cpp.results:
+			i++
+			if err == nil {
+				_, err = cpp.pbp.Print(printTo)
+			}
+			if err != nil {
+				return err
```

Comment:
Ah right, great point! I had this check in one of the versions (that's why
I noticed that allDone weirdness) but guess it fell out among the
refactoring

Luca Bruno <notifications@github.com> schrieb am So., 29. Mai 2016, 21:58:

> In progressutil/iocopy.go
> <https://github.com/coreos/pkg/pull/63#discussion_r65010551>:
>
> > -		if allDone && len(readers) > 0 {
> > -			// We still need to check for errors again, as one may
> > -			// have occurred between when we first checked and when
> > -			// the readers were marked as done
> > -			cpp.lock.Lock()
> > -			errors = cpp.errors
> > -			cpp.lock.Unlock()
> > -			if len(errors) > 0 {
> > -				return errors[0]
> > +		case err := <-cpp.results:
> > +			i++
> > +			if err == nil {
> > +				_, err = cpp.pbp.Print(printTo)
> > +			}
> > +			if err != nil {
> > +				return err
>
> It's only discarding if the copy error was nil...
>
> You are right, I misread it.
>
> I'm not sure about the multiple thing - can you elaborate?
>
> Print boolean signals if all copies have finished. This is ignored here,
> so even if all copies end within the first call, Print will be anyway
> called n times. (I am not sure what the observable behavior will be, I
> guess printing completions result n times).
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/coreos/pkg/pull/63/files/f24e4f6a55a09ae816d44b6ed1c1e09615ffcc73..08ada9715da766febf54352137eb1fa2fbbecf86#r65010551>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ACewN6kc3G1eZzLdDShp5D2BecU_P9ccks5qGe_8gaJpZM4IpSVb>
> .
>